File: app/converters/gemini_to_anthropic.py
"""Convert Gemini format to Anthropic format."""

import json
from typing import Any

from app.models import anthropic as ant
from app.models import gemini as gem


def convert_request(req: gem.GenerateContentRequest) -> ant.MessagesRequest:
    """Convert Gemini generate content request to Anthropic messages request."""
    # System instruction
    system = None
    if req.systemInstruction:
        system = "".join(p.text for p in req.systemInstruction.parts)

    # Convert contents
    messages: list[ant.Message] = []

    for content in req.contents:
        role = "user" if content.role == "user" else "assistant"

        if role == "user":
            content_blocks: list[ant.ContentBlock] = []
            for part in content.parts:
                if isinstance(part, gem.TextPart):
                    content_blocks.append(ant.TextContent(text=part.text))
                elif isinstance(part, gem.FunctionResponsePart):
                    fr = part.functionResponse
                    content_blocks.append(
                        ant.ToolResultContent(
                            tool_use_id=f"toolu_{fr.get('name', '')}",
                            content=json.dumps(fr.get("response", {})),
                        )
                    )

            if content_blocks:
                # If only text, use string content
                if len(content_blocks) == 1 and isinstance(
                    content_blocks[0], ant.TextContent
                ):
                    messages.append(
                        ant.Message(role="user", content=content_blocks[0].text)
                    )
                else:
                    messages.append(ant.Message(role="user", content=content_blocks))

        else:  # assistant
            content_blocks = []
            for i, part in enumerate(content.parts):
                if isinstance(part, gem.TextPart):
                    content_blocks.append(ant.TextContent(text=part.text))
                elif isinstance(part, gem.FunctionCallPart):
                    fc = part.functionCall
                    content_blocks.append(
                        ant.ToolUseContent(
                            id=f"toolu_{i}",
                            name=fc.get("name", ""),
                            input=fc.get("args", {}),
                        )
                    )

            if content_blocks:
                if len(content_blocks) == 1 and isinstance(
                    content_blocks[0], ant.TextContent
                ):
                    messages.append(
                        ant.Message(role="assistant", content=content_blocks[0].text)
                    )
                else:
                    messages.append(
                        ant.Message(role="assistant", content=content_blocks)
                    )

    # Convert tools
    tools = None
    if req.tools:
        tools = []
        for tool_config in req.tools:
            for fd in tool_config.functionDeclarations:
                tools.append(
                    ant.ToolDefinition(
                        name=fd.name,
                        description=fd.description,
                        input_schema=ant.ToolInputSchema(
                            type=fd.parameters.get("type", "object")
                            if fd.parameters
                            else "object",
                            properties=fd.parameters.get("properties", {})
                            if fd.parameters
                            else {},
                            required=fd.parameters.get("required", [])
                            if fd.parameters
                            else [],
                        ),
                    )
                )

    # Generation config
    gen_config = req.generationConfig
    return ant.MessagesRequest(
        model="",  # Will be set by caller
        messages=messages,
        system=system,
        max_tokens=gen_config.maxOutputTokens if gen_config else 4096,
        temperature=gen_config.temperature if gen_config else None,
        top_p=gen_config.topP if gen_config else None,
        top_k=gen_config.topK if gen_config else None,
        stop_sequences=gen_config.stopSequences if gen_config else None,
        tools=tools,
    )


def convert_response(
    resp: ant.MessagesResponse, model: str
) -> gem.GenerateContentResponse:
    """Convert Anthropic messages response to Gemini generate content response."""
    parts: list[gem.Part] = []

    for block in resp.content:
        if isinstance(block, ant.TextContent):
            parts.append(gem.TextPart(text=block.text))
        elif isinstance(block, ant.ToolUseContent):
            parts.append(
                gem.FunctionCallPart(
                    functionCall={
                        "name": block.name,
                        "args": block.input,
                    }
                )
            )

    # Map stop reason
    stop_reason_map = {
        "end_turn": "STOP",
        "stop_sequence": "STOP",
        "max_tokens": "MAX_TOKENS",
        "tool_use": "STOP",
    }
    finish_reason = stop_reason_map.get(resp.stop_reason or "end_turn", "STOP")

    return gem.GenerateContentResponse(
        candidates=[
            gem.Candidate(
                content=gem.ResponseContent(parts=parts if parts else [gem.TextPart(text="")]),
                finishReason=finish_reason,
                index=0,
            )
        ],
        usageMetadata=gem.UsageMetadata(
            promptTokenCount=resp.usage.input_tokens,
            candidatesTokenCount=resp.usage.output_tokens,
            totalTokenCount=resp.usage.input_tokens + resp.usage.output_tokens,
        ),
    )


def convert_stream_event(event: dict[str, Any]) -> dict[str, Any] | None:
    """Convert Anthropic stream event to Gemini stream chunk.

    Returns dict or None if event should be skipped.
    """
    event_type = event.get("type")

    if event_type == "content_block_delta":
        delta = event.get("delta", {})
        if delta.get("type") == "text_delta":
            return {
                "candidates": [
                    {
                        "content": {
                            "parts": [{"text": delta.get("text", "")}],
                            "role": "model",
                        },
                        "index": 0,
                    }
                ]
            }

    elif event_type == "message_delta":
        delta = event.get("delta", {})
        stop_reason = delta.get("stop_reason")
        if stop_reason:
            stop_reason_map = {
                "end_turn": "STOP",
                "stop_sequence": "STOP",
                "max_tokens": "MAX_TOKENS",
                "tool_use": "STOP",
            }
            return {
                "candidates": [
                    {
                        "content": {"parts": [], "role": "model"},
                        "finishReason": stop_reason_map.get(stop_reason, "STOP"),
                        "index": 0,
                    }
                ]
            }

    return None

File: go/internal/converters/gemini_to_anthropic.go
package converters

import (
	"encoding/json"

	"ai_gateway/internal/models"
)

// GeminiToAnthropicRequest converts a Gemini request to Anthropic format
func GeminiToAnthropicRequest(req *models.GenerateContentRequest, model string) (*models.MessagesRequest, error) {
	anthropicReq := &models.MessagesRequest{
		Model:     model,
		MaxTokens: 4096, // Default
	}

	// Convert generation config
	if req.GenerationConfig != nil {
		anthropicReq.Temperature = req.GenerationConfig.Temperature
		anthropicReq.TopP = req.GenerationConfig.TopP
		anthropicReq.TopK = req.GenerationConfig.TopK
		if req.GenerationConfig.MaxOutputTokens != nil {
			anthropicReq.MaxTokens = *req.GenerationConfig.MaxOutputTokens
		}
		anthropicReq.StopSequences = req.GenerationConfig.StopSequences
	}

	// Convert system instruction
	if req.SystemInstruction != nil && len(req.SystemInstruction.Parts) > 0 {
		var systemText string
		for _, part := range req.SystemInstruction.Parts {
			systemText += part.Text
		}
		if systemText != "" {
			anthropicReq.System = systemText
		}
	}

	// Convert contents to messages
	var messages []models.AnthropicMessage
	for _, content := range req.Contents {
		msg := models.AnthropicMessage{}

		// Map role
		switch content.Role {
		case "model":
			msg.Role = "assistant"
		default:
			msg.Role = "user"
		}

		var contentBlocks []models.ContentBlock
		for _, part := range content.Parts {
			if part.Text != "" {
				contentBlocks = append(contentBlocks, models.ContentBlock{
					Type: "text",
					Text: part.Text,
				})
			}
			if part.FunctionCall != nil {
				contentBlocks = append(contentBlocks, models.ContentBlock{
					Type:  "tool_use",
					ID:    generateToolCallID(len(contentBlocks)),
					Name:  part.FunctionCall.Name,
					Input: part.FunctionCall.Args,
				})
			}
			if part.FunctionResponse != nil {
				contentBlocks = append(contentBlocks, models.ContentBlock{
					Type:      "tool_result",
					ToolUseID: generateToolCallID(0), // Gemini doesn't have IDs
					Content:   part.FunctionResponse.Response,
				})
			}
			if part.InlineData != nil {
				contentBlocks = append(contentBlocks, models.ContentBlock{
					Type: "image",
					Source: &models.ImageSource{
						Type:      "base64",
						MediaType: part.InlineData.MimeType,
						Data:      part.InlineData.Data,
					},
				})
			}
		}

		if len(contentBlocks) > 0 {
			msg.Content = contentBlocks
			messages = append(messages, msg)
		}
	}
	anthropicReq.Messages = messages

	// Convert tools
	if len(req.Tools) > 0 {
		var tools []models.AnthropicTool
		for _, tool := range req.Tools {
			for _, decl := range tool.FunctionDeclarations {
				tools = append(tools, models.AnthropicTool{
					Name:        decl.Name,
					Description: decl.Description,
					InputSchema: decl.Parameters,
				})
			}
		}
		anthropicReq.Tools = tools
	}

	return anthropicReq, nil
}

// AnthropicToGeminiResponse converts an Anthropic response to Gemini format
func AnthropicToGeminiResponse(resp map[string]interface{}) (*models.GenerateContentResponse, error) {
	geminiResp := &models.GenerateContentResponse{}

	content, ok := resp["content"].([]interface{})
	if !ok {
		return geminiResp, nil
	}

	var parts []models.GeminiPart
	for _, block := range content {
		blockMap := block.(map[string]interface{})
		blockType := getString(blockMap, "type")

		switch blockType {
		case "text":
			parts = append(parts, models.GeminiPart{
				Text: getString(blockMap, "text"),
			})
		case "tool_use":
			args, _ := blockMap["input"].(map[string]interface{})
			parts = append(parts, models.GeminiPart{
				FunctionCall: &models.GeminiFunctionCall{
					Name: getString(blockMap, "name"),
					Args: args,
				},
			})
		}
	}

	// Convert stop reason
	var finishReason string
	if stopReason, ok := resp["stop_reason"].(string); ok {
		switch stopReason {
		case "end_turn":
			finishReason = "STOP"
		case "max_tokens":
			finishReason = "MAX_TOKENS"
		case "tool_use":
			finishReason = "STOP"
		default:
			finishReason = "STOP"
		}
	}

	geminiResp.Candidates = []models.Candidate{{
		Content: &models.GeminiContent{
			Role:  "model",
			Parts: parts,
		},
		FinishReason: finishReason,
		Index:        0,
	}}

	// Convert usage
	if usage, ok := resp["usage"].(map[string]interface{}); ok {
		inputTokens := getInt(usage, "input_tokens")
		outputTokens := getInt(usage, "output_tokens")
		geminiResp.UsageMetadata = &models.UsageMetadata{
			PromptTokenCount:     inputTokens,
			CandidatesTokenCount: outputTokens,
			TotalTokenCount:      inputTokens + outputTokens,
		}
	}

	return geminiResp, nil
}

// AnthropicStreamToGeminiStream converts an Anthropic stream event to Gemini format
func AnthropicStreamToGeminiStream(eventType string, data map[string]interface{}) ([]byte, error) {
	switch eventType {
	case "content_block_delta":
		delta := data["delta"].(map[string]interface{})
		deltaType := getString(delta, "type")

		if deltaType == "text_delta" {
			text := getString(delta, "text")
			resp := models.GenerateContentResponse{
				Candidates: []models.Candidate{{
					Content: &models.GeminiContent{
						Role:  "model",
						Parts: []models.GeminiPart{{Text: text}},
					},
					Index: 0,
				}},
			}
			return json.Marshal(resp)
		}

	case "message_delta":
		delta := data["delta"].(map[string]interface{})
		stopReason := getString(delta, "stop_reason")

		var finishReason string
		switch stopReason {
		case "end_turn":
			finishReason = "STOP"
		case "max_tokens":
			finishReason = "MAX_TOKENS"
		default:
			finishReason = "STOP"
		}

		resp := models.GenerateContentResponse{
			Candidates: []models.Candidate{{
				Content: &models.GeminiContent{
					Role:  "model",
					Parts: []models.GeminiPart{},
				},
				FinishReason: finishReason,
				Index:        0,
			}},
		}
		return json.Marshal(resp)
	}

	return nil, nil
}
